Chore: Update Slate to 0.47.8

kaydelaney opened the pull request:
**What this PR does / why we need it**:
Updates slate to 0.47.8 and refactors code making use of slate accordingly.

**Which issue(s) this PR fixes**:
Closes #17430

kaydelaney committed:
Chore: Update Slate to 0.47.8
Closes #17430

kaydelaney committed:
Add slate and immutable to grafana-ui deps

kaydelaney committed:
Fixes some small regressions introduced

davkal committed:
Fix suggestions for multiple query fields

davkal committed:
Pin upgraded dependencies

davkal committed:
Prettier fix

kaydelaney committed:
Remove original slate-react dependency

kaydelaney committed:
Fix tiny-invariant dep

kaydelaney committed:
(Temporarily) comments out source maps for grafana-ui

kaydelaney committed:
Editor: Fixes problems with azure and datalinks editors caused by slate upgrade

davkal reviewed (approved):
LGTM, tested Prometheus, Loki, Azure Logs, Data links

@dprokop do you want to take another look too?

hugohaggmark reviewed (commented):
Wow, what a huge undertaking! Great job with this.
I left some small comments and I found an unexpected behaviour when you select all the text using SHIFT+OPTION+LEFT arrow on MacOS all the text is first selected and the automatically erased.

![unexpected-select-all](https://user-images.githubusercontent.com/562238/65314332-6ce47680-db96-11e9-8b7d-468fc2154288.gif)

dprokop commented on packages/grafana-ui/rollup.config.ts:
```diff
@@ -47,19 +47,20 @@ const buildCjsPackage = ({ env }) => {
       }),
       resolve(),
-      sourceMaps(),
+      // sourceMaps(),
```
Let's leave it here for now and add a comment why this is actually commented out. I need to investigate it deeper at some point. Probably worth creating an issue too.

kaydelaney commented:
@hugohaggmark Thanks for your review! Will look into that  unexpected behavior you mentioned.

dprokop commented:
@kaydelaney minor UX observation in DataLinks - pressing Backspace closes suggestions menu but does not remove character from the input

dprokop commented on public/app/features/explore/slate-plugins/clear.test.tsx:
```diff
@@ -0,0 +1,42 @@
+import Plain from 'slate-plain-serializer';
+import React from 'react';
+import { Editor } from '@grafana/slate-react';
```
What's the plan for maintaining this vs. eventually going back to non-forked version?

kaydelaney replied:
I think we can switch back to the non-forked version as soon as my fixes get merged, but unfortunately things move fairly slowly upstream

dprokop reviewed (approved):
Really solid and impressive work @kaydelaney!

kaydelaney committed:
Fixes several small remaining issues
- Fixes issue where selection shortcuts would result in text getting deleted
- Fixes issue in DataLinks input where backspace didn't delete text while typeahead menu was open
- Reduces scope of selection shortcuts plugin, since a lot of behavior "just works" with new slate
- Slightly simpligies positioning logic for typeahead

kaydelaney commented:
Fixed the issues mentioned by @hugohaggmark and @dprokop. Should be okay to merge! Thoughts @davkal?

torkelo commented:
great work on this tricky upgrade! Hopefully, next upgrade won't require as much work :)
